### PR TITLE
Function that tests whether a 3D point is inside a mesh.

### PR DESCRIPTION
**Pull request by tinevez**
Because we use the perturbation in Z trick, it is not strictly an interior test. 
It tries to have decent performance, without using AABB meshes, by using an fast face retrieval:
- before running location query:
  - the triangles are sorted  by minZ
  - separately they sorted by maxZ
- for each point:
  - we retrieve the triangles that share its Z position from the sorted arrays above
  - we cast a ray from the point towards positive X, then look for how many times we cross a triangle from the subset generated above
   - we return inside or outside depending on the number of intersections

This algo can deal with border cases where the ray crosses an edge between two triangles that may or may not correspond to crossing the mesh.

Its absolute performance was not judged as fantastic by Ella with 0.2ms / point :) but we don't know yet how other implementations fare. 
https://imagesc.zulipchat.com/#narrow/stream/391996-.5B2023-06.5D-scenery.2Bsciview-hackathon-dresden/topic/Half-edge.20data.20structure.20in.20Java

### CODE REVIEW

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+			final long v0 = triangles.vertex0( t );
+			final long v1 = triangles.vertex1( t );
+			final long v2 = triangles.vertex2( t );
+
+			final double minZ = minZ( vertices, v0, v1, v2, eps );
+			minZs[ t ] = minZ;
+			final double maxZ = maxZ( vertices, v0, v1, v2, eps );
+			maxZs[ t ] = maxZ;
+		}
+		this.indexMin = SortArray.quicksort( minZs );
```
Comment (ctrueden) [inline src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:112]: It's crazy to me that we have our own quicksort implementation in this library, just to get around issues of sorting indices vs values, and boxing/unboxing of ints. I did some digging to see what libraries exist providing this functionality without a custom sort implementation. E.g. something with an `IntComparator` that has signature `int compareTo(int a, int b)` and then sort routines that operate via such a comparator object. But none of Trove, Guava, or Apache Commons provides this. Only Fastutil, as far as I could find, but the fastutil-core JAR is already >6MB. :confused: 

Is it worth a fastutil dependency to get rid of this custom implementation? At least fastutil doesn't have any dependencies of its own...

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+		this.indices = new int[ indexMin.length ];
+		this.intersection = new double[ 3 ];
+	}
+
+	public boolean isInside( final RealPoint p )
+	{
```
Comment (ctrueden) [inline src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:118]: Shall we raise an `IllegalArgumentException` if `p.numDimensions() != 3`? Or `< 3`?

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+		 *  |                               |
+		 *  |   X --------> .               |
+		 *  |              / \              |
+		 *  |_____________/   \_____________|
+		 * 
+		 * </pre>
+		 * 
+		 * The X point is inside, but the ray-casting will cross the edge
+		 * protruding from the bottom, exactly on the line. In that case this
+		 * should not count as one crossing of the bounds.
```
Comment (ctrueden) [inline src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:199]: I'm confused by this explanation. The issue is that the "interior" (really boundary) point lies at exactly the vertex shared by two edge triangles there? Will the ray not then cross both triangles, and thus count as two crossings, and therefore not flip the algorithm's even-odd criterion? Probably I just do not understand the algorithm correctly... or I misinterpret the drawing...

Furthermore, as stated below, this algorithm uses the shifted position trick to avoid singularities... so we shouldn't have rays hitting exactly through vertices or edges or faces anyway, should we?

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+		double previousX = Double.NaN;
+		double previousN = xNormals.get( 0 );
+		for ( int i = 0; i < xIntersect.size(); i++ )
+		{
+			final double v = xIntersect.get( i );
+			final double n = xNormals.get( i );
+
+			if ( ( v != previousX ) || ( n * previousN > 0 ) )
+				nCross++;
```
Comment (ctrueden) [inline src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:213]: Could be simplified to `inside = !inside` with `boolean inside` instead of `int nCross`. Then if there are more than 2^31 crossings we won't overflow :sweat_smile: (of course I'm joking, the `xIntersect` array can't get that large anyway)

**Review by ctrueden — APPROVED** [review]
I made some minor comments. I started digging into the algorithm, but I think it probably does not make sense to scrutinize it too intensely at the moment. The initial tests provide proof that it fundamentally works (although a couple more nuanced tests would be a good idea). The only change I would advise making immediately is renaming from `InteriorPointTest` to something without `Test` in the name, perhaps just `Interior` or maybe `MeshInterior`.

I started talking with ChatGPT to revive my attempt to use CGAL to do this sort of test. In a nutshell: with AABB trees and BVHs. I wish I could find the time to get it working in CGAL with these structures and then benchmark the performance compared to this Z-slice-based implementation, but realistically probably not any time soon, unless we find a critical bug in this algorithm and decide we really need it more urgently.

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
+package net.imglib2.mesh.alg;
+
+import static net.imglib2.mesh.util.MeshUtil.maxZ;
+import static net.imglib2.mesh.util.MeshUtil.minZ;
+import static net.imglib2.mesh.util.MeshUtil.mround;
+
+import java.util.Arrays;
+
+import gnu.trove.list.array.TDoubleArrayList;
```
[the remark this answers is not in the document]
Comment (tinevez) [reply src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:9]: Let's do this after #2 and #3 are merged.

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+		/**
+		 * Only consider intersections that are unique. If we have twice the
+		 * same X intersection, then it means we crossed an edge. In that case
+		 * we need to test whether the 2 triangles of this edge are facing the
+		 * same direction. If yes (their normals along x have the same sign), it
+		 * means that we are crossing the mesh boundary. If not, they do not
+		 * count as crossing the boundary.
+		 * 
+		 * Fantastic drawing that illustrates this situation:
```
[the remark this answers is not in the document]
Comment (tinevez) [reply src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:185]: That's the power of a Madison hackathon. It fosters the artist in me.

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+import net.imglib2.mesh.Meshes;
+import net.imglib2.mesh.obj.Mesh;
+import net.imglib2.mesh.obj.Triangles;
+import net.imglib2.mesh.obj.Vertices;
+import net.imglib2.mesh.util.MeshUtil;
+import net.imglib2.mesh.util.SortArray;
+import net.imglib2.mesh.util.SortBy;
+import net.imglib2.util.Intervals;
+
+public class InteriorPointTest
```
[the remark this answers is not in the document]
Comment (tinevez) [reply src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:21]: Good idea!

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java (hunk 1 above)
Comment (tinevez) [reply src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:112]: As above, let's review this after we thinned the dependencies.

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java (hunk 3 above)
Comment (tinevez) [reply src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:199]: Alors, the issue is not whether the ray crosses a *vertex*, which is not a problem here, as you noticed.
The issue is when a ray crosses an *edge*. This is no guarantee that an edge will not be crossed by a ray, even if we shift the Z position of vertices. In that case crossing the edge will mean different thing whether the 'angle' of the triangles adjacent to the edge.
It was very hard (for me) to conceptualize, fix, and build a test case for this.

File: src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java (hunk 4 above)
Comment (tinevez) [reply src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java:213]: It was useful when debugging. I am sure we will debug again. Let's leave it.

**Comment by tinevez** [comment]
Done!
Careful again, I git-push-forced.
